addons/mail/static/src/messaging/entity/entity/entity.js

```diff
@@ -0,0 +1,824 @@
+         * @param {Object} param1
+         * @param {string} param1.relationName
+         */
+        static __unlinkSingleOne2One(entity, { relationName }) {
+            if (!this.get(entity)) {
+                // Entity has already been deleted.
+                // (e.g. unlinking one of its reverse relation was causal)
+                return;
+            }
+            const relation = this.relations[relationName];
+            const storedRelationName = this.__getStoredRelationName(relationName);
+            const prevValue = entity[storedRelationName];
+            entity._write({ [storedRelationName]: undefined });
+            const RelatedEntity = this.env.entities[relation.to];
+            if (prevValue) {
+                const related = RelatedEntity.get(prevValue);
+                const storedRelatedRelationName =
+                    RelatedEntity.__getStoredRelationName(relation.inverse);
+                related._write({ [storedRelatedRelationName]: undefined });
+            }
+        }
+
+        /**
+         * @static
+         * @private
+         * @param {mail.messaging.entity.Entity} entity
+         * @param {Object} [data={}]
+         */
+        static __update(entity, data = {}) {
+            entity._update(data);
+        }
+
+    }
+
+    Object.assign(Entity, {
+        /**
+         * Registry containing data to make Entity classes. Entity classes should
+         * register themselves though static method `Entity.registerNewEntity()`.
+         *
+         * Format:
+         *
+         *   <relation-name>:
+         *      {
+         *         Factory: function that produce Entity class.
+         *         plugins: list of extensions to apply on the Entity class.
+         *                  Each entity classes should itself be designed to process
+         *                  these plugins.
+         *      }
+         */
+        registry: {},
+        /**
+         * Schema of relations for this entity.
+         *
+         * Format:
+         *
+         *   <relation-name>:
+         *      {
+         *         inverse: Name of inverse relation on related entity.
+         *         isCausal: boolean that determines whether the related entities
+         *                   are deeply connected to existence of current entity.
+         *                   (default: false)
+         *         to: Name of the related entity. Just for documentation sake.
+         *         type: Type of the relation on this entity.
+         *               Either 'one2one', 'one2many', 'many2one' or 'many2many'.
+         *      }
+         */
+        relations: {
+            /**
+             * Related dialog of entity when dialog content is directly linked to
+             * an entity that models a UI component, such as AttachmentViewer. Such
+             * entities must be created from @see `Dialog.open()` and cannot be
+             * singleton entities.
+             */
+            dialog: {
```

Comment:
Why is dialog here?!

Reply:
this is to link a dialog to an UI entity, like attachment viewer, thread viewer, etc.

I was thinking of improving this later by finding a way to partition entities in 2 groups (maybe ui/model entities?)